### ISSUE TITLE
[question] fully self hosting packages from conan center on local artifactory

### ISSUE THREAD

**Issue by dornbirndevelops** [conan-io/conan#9436]
Within a private network, I have set up an instance of Artifactory that provides some packages.
These packages sometimes require official libraries (boost as an example).
Due too strict firewall rules (that I cannot really change) I unfortunately cannot use the official conan servers as remotes.
The only remotes available for conan clients within this private network are repositories of this local Artifactory instance.

How can I as a network administrator synchronize a selection of these packages available on conan center onto this local Artifactory instance? By _sync_ I mean everything related to build the package in this private network, e.g. the receipe, the sources, the patches. Is there a way conan can help here?
Or do I need to manually sync the repository, self host all sources in conandata.yml and patch those acordingly?

I added the question here because I think the problem can be also seen from a more general way (source x, target y).
If I'm in the wrong repository here, feel free to let me know.

- [x] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).

**Comment by memsharded** [comment]
Hi @dornbirndevelops 

For Conan packages, you could proxy ConanCenter with an Artifactory remote repo, and clients will only connect to your Artifactory. But that doesn't bring automatically the full sources stored in other servers, like Github. If you need everything reference from ConanCenter under your own infrastructure, yes, the way would probably be:

- Fork conan-center-index recipes repo
- Get the sources stored in ``conandata.yml`` files. Upload them to an Artifactory generic repo.
- Replace the ``conandata.yml`` url with the new one pointing to Artifactory generic repo.
- Create and upload the package to Artifactory.

The good thing is that this should be relatively simple to automate, as the information is structured in ``conandata.yml`` and along the conan-center-index repo, so keeping an ordered list of the specific packages and versions that you want to create from the repo could feed this automation.

**Comment by memsharded** [comment]
Note for team: It would be great to have this documented and shared in conan-center-index repo.

**Comment by dornbirndevelops** [comment]
Hi @memsharded 

I managed to accomplish my goal without the need to fork the official repository.
Instead I wrote a small hook for `pre_upload_recipe` that manipulates the mirrors in the downloaded conandata.yml before its being uploaded to the target remote. Additional steps that were needed was to repackage the extracted sources and update the manifest hashes.
With this hook in place the usage is as simple as:

```
conan download -re <reference> -r conancenter
conan upload <reference> -r <remote_target>
```

**Comment by memsharded** [comment]
Thanks for sharing your approach @dornbirndevelops !

Just a word of caution: you are modifying what are supposed to be immutable packages. This will probably not work correctly if using revisions, because the revisions themselves would need to be recomputed, and that can produce changes in the package-ids downstream, etc, etc. Please take that Conan 2.0 will be using revisions always, and things like the cache 2.0 have already been designed and implemented with this immutability in mind.

**Comment by dornbirndevelops** [comment]
I just figured out that the conan-center-index repository is actually really simple to export.
Also for better transparency and less "side-effect" code, I will try and go with the [solution](https://github.com/conan-io/conan/issues/9436#issuecomment-902306840) described here.
Thanks again 👍🏼

**Comment by SeanSnyders** [comment]
Has this process been documented yet as per https://github.com/conan-io/conan/issues/9436#issuecomment-902307219?
![image](https://user-images.githubusercontent.com/1384135/157367439-2d856f3f-fb8d-473e-8b8f-7c65dbac8eb2.png)

**Comment by memsharded** [comment]
Created issue https://github.com/conan-io/conan-center-index/issues/9752 in conan-center-index repo for this.

**Comment by AndreyMlashkin** [comment]
> Hi @memsharded
> 
> I managed to accomplish my goal without the need to fork the official repository. Instead I wrote a small hook for `pre_upload_recipe` that manipulates the mirrors in the downloaded conandata.yml before its being uploaded to the target remote. Additional steps that were needed was to repackage the extracted sources and update the manifest hashes. With this hook in place the usage is as simple as:
> 
> ```
> conan download -re <reference> -r conancenter
> conan upload <reference> -r <remote_target>
> ```

@dornbirndevelops could you maybe share your automation script for this?

**Comment by memsharded** [comment]
Update: The automatic "backup sources" feature is already merged and released in Conan 2.0.9, and it is already running in prod in ConanCenter. It will be publicly launched soon, the PR with the docs is https://github.com/conan-io/docs/pull/3365, and there will be a blog post. 

This will avoid modifying the ``conandata.yml`` completely, and still being able to fully self-host packages sources in your own server.
